lib/application_group.ts

```diff
@@ -0,0 +1,224 @@
+           * create a new blank slate app */
+        const descriptors = this.getDescriptors();
+        if (descriptors.length === 0) {
+          return this.addNewApplication();
+        } else {
+          return this.loadApplicationForDescriptor(descriptors[0]);
+        }
+      }
+    } else if (source === DeinitSource.Lock && sideffects) {
+      /** Recreate the same application from scratch */
+      const descriptor = this.descriptorForApplication(application);
+      return this.loadApplicationForDescriptor(descriptor);
+    }
+  }
+
+  /**
+   * Notifies observer when the primary application has changed.
+   * Any application which is no longer active is destroyed, and
+   * must be removed from the interface.
+   */
+  public addApplicationChangeObserver(callback: AppGroupChangeCallback) {
+    this.changeObservers.push(callback);
+    if (this.primaryApplication) {
+      callback();
+    }
+    return () => {
+      removeFromArray(this.changeObservers, callback);
+    }
+  }
+
+  private notifyObserversOfAppChange() {
+    for (const observer of this.changeObservers) {
+      observer();
+    }
+  }
+
+  public async setPrimaryApplication(application: SNApplication, persist = true) {
+    if(this.primaryApplication === application) {
+      return;
+    }
+    if (!this.applications.includes(application)) {
+      throw Error('Application must be inserted before attempting to switch to it');
+    }
+    if (this.primaryApplication) {
+      this.primaryApplication.deinit(DeinitSource.AppGroupUnload);
+    }
+    this.primaryApplication = application;
+    const descriptor = this.descriptorForApplication(application);
+    this.setDescriptorAsPrimary(descriptor);
+    this.notifyObserversOfAppChange();
+    if (persist) {
+      await this.persistDescriptors();
+    }
+  }
+
+  setDescriptorAsPrimary(primaryDescriptor: ApplicationDescriptor) {
+    for(const descriptor of this.getDescriptors()) {
+      descriptor.primary = descriptor === primaryDescriptor;
+    }
+  }
+
+  private async persistDescriptors() {
+    this.deviceInterface!.setRawStorageValue(
+      RawStorageKey.DescriptorRecord,
+      JSON.stringify(this.descriptorRecord)
+    );
+  }
+
+  public async renameDescriptor(descriptor: ApplicationDescriptor, label: string) {
+    descriptor.label = label;
+    await this.persistDescriptors();
+  }
+
+  public removeDescriptor(descriptor: ApplicationDescriptor) {
+    delete this.descriptorRecord[descriptor.identifier];
+    return this.persistDescriptors();
+  }
+
+  private descriptorForApplication(application: SNApplication) {
+    return this.descriptorRecord[application.identifier];
+  }
+
+  public async addNewApplication(label?: string) {
+    const identifier = await Uuid.GenerateUuid();
+    const index = this.getDescriptors().length + 1;
+    const descriptor = {
+      identifier: identifier,
+      label: label || `Application ${index}`
+    } as ApplicationDescriptor;
```

Comment:
should be
```ts
const descriptor: ApplicationDescriptor = { ... }
```
Instead of casting. That way if we ever change ApplicationDescriptor we'll have errors pointing to places where we need to update code